File: cpp/example_code/cloudwatch/cwe_put_events.cpp
/*
   Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
#include <aws/core/Aws.h>

#include <aws/events/CloudWatchEventsClient.h>
#include <aws/events/model/PutEventsRequest.h>
#include <aws/events/model/PutEventsResult.h>

#include <aws/core/utils/Outcome.h>

#include <iostream>

Aws::String MakeDetails(const Aws::String &key, const Aws::String& value)
{
    Aws::Utils::Json::JsonValue valueEntry;
    valueEntry.AsString(value);

    Aws::Utils::Json::JsonValue detailMap;
    detailMap.WithObject(key, valueEntry);

    return detailMap.WriteReadable();
}

/**
 * Posts a sample cloudwatch event, based on command line input
 */
int main(int argc, char** argv)
{
    if (argc != 4)
    {
        std::cout << "Usage: cwe_put_events <resource_arn> <sample_key> <sample_value>" << std::endl;
        return 1;
    }

    Aws::String resourceArn(argv[1]);
    Aws::String eventKey(argv[2]);
    Aws::String eventValue(argv[3]);

    Aws::SDKOptions options;
    Aws::InitAPI(options);

    {
        Aws::CloudWatchEvents::CloudWatchEventsClient cwe_client;

        Aws::CloudWatchEvents::Model::PutEventsRequestEntry eventEntry;
        eventEntry.SetDetail(MakeDetails(eventKey, eventValue));
        eventEntry.SetDetailType("sampleSubmitted");
        eventEntry.AddResources(resourceArn);
        eventEntry.SetSource("aws-sdk-cpp-cloudwatch-example");

        Aws::CloudWatchEvents::Model::PutEventsRequest putEventsRequest;
        putEventsRequest.AddEntries(eventEntry);

        auto putEventsOutcome = cwe_client.PutEvents(putEventsRequest);
        if (!putEventsOutcome.IsSuccess())
        {
            std::cout << "Failed to post cloudwatch event: " << putEventsOutcome.GetError().GetMessage() << std::endl;
        }
        else
        {
            std::cout << "Successfully posted cloudwatch event" << std::endl;
        }
    }

    Aws::ShutdownAPI(options);

    return 0;
}

File: javascript/example_code/cloudwatch/cwe_putevents.js
/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/

//snippet-sourcedescription:[cwe_putevents.js demonstrates how to send custom events to Amazon CloudWatch Events so they can be matched with rules.]
//snippet-service:[cw_events]
//snippet-keyword:[JavaScript]
//snippet-sourcesyntax:[javascript]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon CloudWatch Events]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[2018-06-02]
//snippet-sourceauthor:[AWS-JSDG]

// ABOUT THIS NODE.JS SAMPLE: This sample is part of the SDK for JavaScript Developer Guide topic at
// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/cloudwatch-examples-sending-events.html 

// snippet-start:[cwEvents.JavaScript.cwe.putEvents]
// Load the AWS SDK for Node.js
var AWS = require('aws-sdk');
// Set the region 
AWS.config.update({region: 'REGION'});

// Create CloudWatchEvents service object
var cwevents = new AWS.CloudWatchEvents({apiVersion: '2015-10-07'});

var params = {
  Entries: [
    {
      Detail: '{ \"key1\": \"value1\", \"key2\": \"value2\" }',
      DetailType: 'appRequestSubmitted',
      Resources: [
        'RESOURCE_ARN',
      ],
      Source: 'com.company.app'
    }
  ]
};

cwevents.putEvents(params, function(err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    console.log("Success", data.Entries);
  }
});
// snippet-end:[cwEvents.JavaScript.cwe.putEvents]
